src/main/java/com/r307/arbitrader/service/TickerService.java

```diff
@@ -33,9 +34,9 @@
     private final ExchangeService exchangeService;
     private final ErrorCollectorService errorCollectorService;
 
-    Map<String, Ticker> allTickers = new HashMap<>();
-    List<TradeCombination> pollingExchangeTradeCombinations = new ArrayList<>();
-    List<TradeCombination> streamingExchangeTradeCombinations = new ArrayList<>();
+    private final Map<String, Ticker> allTickers = new HashMap<>();
+    private final Map<Exchange, Set<TradeCombination>> pollingExchangeTradeCombination = new HashMap<>();
+    private final Map<Exchange, Set<TradeCombination>> streamingExchangeTradeCombination = new HashMap<>();
```

Comment:
I think I've figured out why these make me uncomfortable. I had doubts before but couldn't figure out how to explain them.

The idea behind `TickerStrategy` was to hide the implementation so that you didn't need to know _how_ it got the tickers and you didn't need to care whether an exchange was polling or streaming. `TradeCombination` shouldn't know or care about streaming vs. polling - it's part of the trade analysis layer, not the ticker fetch layer.

The `StreamingTickerStrategy` has the correct implementation; it just fetches the ticker and publishes the event.

The problem starts in the `StreamingTickerEventListener` where it has knowledge of whether the exchange is streaming or not and the trading process is different based on streaming or not. The `TradingService` shouldn't know or care about streaming. It just gets the tickers stored in `TickerService` and trusts that they are the most up to date. I understand this as being an optimization to only check exchanges that might have changed due to an event and skip the polling ones, but it breaks the encapsulation and even then I think it's incorrect. The polling trade combinations are updated in a different thread and actually could have changed. 

A better solution is what we talked about earlier today. Having the exchange in the event and filtering the list of trade combinations down to only the ones that have that exchange. But, we should do that for everything: the other `TickerStrategy` implementations should publish ticker events too. The event listener should be the only way we trigger the trade logic, and the trade logic should only get tickers from `TickerService`. That leaves the trade logic and ticker fetching isolated from one another.

Once we get to that point it should become easier to split the trade analysis part away from the trade execution part, so that we can run multiple trade analyses in parallel but only allow one trade at a time. I think multiple concurrent trades will follow relatively easily from there.

Reply:
Thanks for the detailed code review. I have changed the structure in the way you recommended. Now tickers are updated on TickerService#allTickers and whenever there is a new ticker update a new TradeAnalysisEvent shoud be fired (still missing the trade analysis service logic).

To get the a ticker one has to call TickerService#getTickers()